architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,793 @@
+
+    UCP -->|"Reads/Writes"| Database["etcd / PostgreSQL"]
+    UCP -.->|"Only holds metadata"| UserApps["User Applications"]
+```
+
+### Detailed Design
+
+```mermaid
+graph TD
+    Start[User runs rad upgrade kubernetes] --> ParseArgs[Parse command arguments]
+    ParseArgs --> ValidateVersion[Validate version compatibility]
+    ValidateVersion --> AcquireLock[Acquire upgrade lock]
+    AcquireLock --> RunPreflights[Run pre-flight checks]
+    RunPreflights --> UserDataBackup[Create user data backup]
+    UserDataBackup --> PlanUpgrade[Calculate upgrade plan]
+    PlanUpgrade --> ExecuteHelmUpgrade[Execute Helm chart upgrade]
+    ExecuteHelmUpgrade --> MonitorProgress[Monitor upgrade progress]
+    MonitorProgress --> VerifyComponents[Verify component health]
+    VerifyComponents --> Success{Successful?}
+    Success -- Yes --> ReleaseLock[Release upgrade lock]
+    Success -- No --> UserDataRestore[Restore user data]
+    UserDataRestore --> RollbackHelm[Rollback Helm changes]
+    RollbackHelm --> ReleaseLock
+    ReleaseLock --> End[Display results to user]
+```
+
+**Version Detection and Validation:**
+
+The upgrade process begins with detecting the currently installed Radius version (the versions of the components in the cluster) and validating it against the requested target version. We can build an interface (or use/improve an existing one) to achieve this:
+
+```go
+// VersionValidator interface
+type VersionValidator interface {
+    ValidateTargetVersion(currentVersion, targetVersion string) error
+    GetLatestVersion() (string, error)
+    IsValidVersion(version string) bool
+}
+```
+
+This interface will be implemented (or existing will be improved) to handle version comparisons, prevent downgrades, and resolve the "latest" version tag to a specific version number. The implementation will (probably) connect to the GitHub API to fetch available release versions when needed.
+
+**Upgrade Lock Mechanism:**
+
+- To prevent concurrent data‐modifying operations during `rad upgrade kubernetes`, we’ll rely exclusively on datastore locks (no Kubernetes leases).
+
+```go
+// UpgradeLock is implemented per datastore (Postgres, etcd) to serialize upgrades.
+type UpgradeLock interface {
+    // AcquireLock blocks until it obtains an exclusive lock or the context deadline is exceeded.
+    AcquireLock(ctx context.Context) error
+
+    // ReleaseLock frees the lock immediately so others can proceed.
+    ReleaseLock(ctx context.Context) error
+
+    // IsUpgradeInProgress returns true if a valid (non‑stale) lock is held by another process.
+    IsUpgradeInProgress(ctx context.Context) (bool, error)
+}
+```
+
+**Timeouts:** callers must supply a context with a finite deadline (e.g. 2 min) to avoid blocking forever.
+**Stale‑lock detection:** each lock has a TTL/heartbeat; expired leases are auto‑cleaned before AcquireLock.
+Force cleanup: --force flag allows manual removal of stale/orphaned locks.
+
+Usage in CLI commands:
```

Comment:
https://github.com/radius-project/design-notes/pull/87#discussion_r2032108184 would still be an issue if the we are checking the lock in the CLI.